### DISCUSSION TITLE
Smart-splits.nvim + Wezterm

### DISCUSSION THREAD

**Discussion by masoniis** [mrjones2014/smart-splits.nvim#52] [Q&A] [answered]
Does anyone else use Wezterm as their terminal emulator alongside this plugin? I noticed that on Wezterm there is a noticeable delay (maybe 500-1000ms) when performing a circular window navigation (and this doesn't appear when I switch to macOS default terminal).

I'm not sure if this is a fixable problem, or if it's just a result of the slowness of Wezterm itself when paired with the circular navigation.

**Comment by mrjones2014** [comment]
I do ✋ 

For me, though, the delay is the same whether it's the circular navigation or not. I think it is just due to the fact that the Wezterm keybinds require getting the foreground process name to detect if vim/nvim is running, which unfortunately can be an expensive operation as mentioned here: https://wezfurlong.org/wezterm/config/lua/PaneInformation.html

I'll keep thinking on if there is any other way to achieve this.

**Reply by mrjones2014** [reply]
Actually, I think that's not what you were talking about. I see a bit of a delay when doing the circular navigation just in Neovim splits, without any Wezterm multiplexer splits.

**Reply by mrjones2014** [reply]
It must still have something to do with the multiplexer integration of Wezterm. Disabling it removes the delay.

**Reply by mrjones2014** [reply]
It's this code that's causing the issue, but I am not sure how else I can implement this.

https://github.com/mrjones2014/smart-splits.nvim/blob/653420ba71b13503cbd92ccdbd123fb629a87e6e/lua/smart-splits/mux/wezterm.lua#L17-L30

If you use `vim.env.WEZTERM_PANE`, you just get the pane ID that Neovim is running inside of, not the actual currently focused pane ID.

**Reply by mrjones2014** [reply]
Maybe @wez will have an idea?

**Reply by mrjones2014** [reply]
I think it is the fact that we have to `vim.json.decode` the output causing a performance bottleneck.

**Reply by wez** [reply]
Does the issue persist with a build from today? https://github.com/wez/wezterm/commit/f57df9bbaf02c81112b8a52fb11e92bd640a31f3 took out some config parsing overhead from the `wezterm cli` subcommands that might possibly be influencing this.

**Reply by mrjones2014** [reply]
Building now to verify but looking at the changes in the commit, I think it will still happen. The command being run in the snippet above is `wezterm cli list-clients --format=json`, and the config is still loaded for `wezterm cli` anyway.

**Reply by mrjones2014** [reply]
Built from latest `main` and the issue still persists.

I think the main issue is that we're having to parse JSON and potentially sort the results as well, both of which can be expensive.

Wezterm doesn't have any way to output _just the current pane ID of the current client_ as a standalone string, correct?

**Reply by wez** [reply]
json parse and sort of a few hundred bytes shouldn't take hundreds of milliseconds, so it's probably not that.  What does `time wezterm cli list-clients --format=json`  show?

**Reply by masoniis** [reply]
`time wezterm cli list-clients --format=json` returns 0.072 total for me. It's possible that this is the time contributing most to the delay and I grossly overestimated the total delay it takes when switching on a circular pane. I suppose it's just a bit jarring when having instantaneous "normal" switches that lead into the longer-delayed circular switch.

Heres a video showing the speed: [video](https://imgur.com/a/IcYrDhz)

It's not too big a deal, perhaps just a tradeoff with using something more feature-rich like wezterm

**Reply by wez** [reply]
Hmm, that's still pretty high, so I whipped up https://github.com/wez/wezterm/commit/084e6569969bddde9ab085ba423ea5e608857bdf brings that down to 3ms.

**Reply by mrjones2014** [reply]
The issue still persists with that commit, I built and tested locally. I think it may also be due to the Wezterm keybinds I'm using to support the integration, I think the docs said that `pane:get_foreground_process_name()` can be expensive?
```lua
local w = require('wezterm')

-- Equivalent to POSIX basename(3)
-- Given "/foo/bar" returns "bar"
-- Given "c:\\foo\\bar" returns "bar"
local function basename(s)
  return string.gsub(s, '(.*[/\\])(.*)', '%2')
end

local function is_vim(pane)
  local process_name = basename(pane:get_foreground_process_name())
  return process_name == 'nvim' or process_name == 'vim'
end

local direction_keys = {
  Left = 'h',
  Down = 'j',
  Up = 'k',
  Right = 'l',
  -- reverse lookup
  h = 'Left',
  j = 'Down',
  k = 'Up',
  l = 'Right',
}

local function split_nav(resize_or_move, key)
  return {
    key = key,
    mods = resize_or_move == 'resize' and 'META' or 'CTRL',
    action = w.action_callback(function(win, pane)
      if is_vim(pane) then
        -- pass the keys through to vim/nvim
        win:perform_action({
          SendKey = { key = key, mods = resize_or_move == 'resize' and 'META' or 'CTRL' },
        }, pane)
      else
        if resize_or_move == 'resize' then
          win:perform_action({ AdjustPaneSize = { direction_keys[key], 3 } }, pane)
        else
          win:perform_action({ ActivatePaneDirection = direction_keys[key] }, pane)
        end
      end
    end),
  }
end

return {
  keys = {
    -- move between split panes
    split_nav('move', 'h'),
    split_nav('move', 'j'),
    split_nav('move', 'k'),
    split_nav('move', 'l'),
    -- resize panes
    split_nav('resize', 'h'),
    split_nav('resize', 'j'),
    split_nav('resize', 'k'),
    split_nav('resize', 'l'),
  },
}
```

**Reply by wez** [reply]
querying process information can have high and highly variable latency. You could try hard-coding your `is_vim` function to always return true (or false) to see if the latency is better to confirm/deny that theory.

**Reply by mrjones2014** [reply]
Hm, that does not fix the issue. However, this change fixes the performance issue, but introduces a bug in the plugin:

```lua
function M.current_pane_id()
  -- local output = wezterm_exec({ 'list-clients', '--format=json' })
  -- local data = vim.json.decode(output)
  -- if #data == 0 then
  --   return nil
  -- end
  -- -- if more than one client, get the active one
  -- if #data > 1 then
  --   table.sort(data, function(a, b)
  --     return a.idle_time.nanos < b.idle_time.nanos
  --   end)
  -- end
  -- return data[1].focused_pane_id
  return vim.env.WEZTERM_PANE
end
```

**Reply by mrjones2014** [reply]
Created #53 from this discussion.

**Reply by masoniis** [reply]
Closing discussion due to fix in [#53](https://github.com/mrjones2014/smart-splits.nvim/issues/53) 🎉

**Comment by mrjones2014** [comment, marked as answer]
As noted in #53, a commit has been made to improve performance, make sure you are using a Wezterm build with this commit included to take advantage of it. It should be included in nightlies starting tomrrow.
https://github.com/wez/wezterm/commit/96f1585b103162b042e2593567884dfcbe32ca21

**Reply by wez** [reply]
wezterm nightlies with that change are already available!

**Reply by mrjones2014** [reply]
Oh nice, thanks!
